packages-ui/roosterjs-react/lib/emoji/plugin/createEmojiPlugin.ts

```diff
@@ -0,0 +1,426 @@
+
+// Regex looks for an emoji right before the : to allow contextual search immediately following an emoji
+// MATCHES: 0: 😃:r
+//          1: 😃
+//          2: :r
+
+const EMOJI_BEFORE_COLON_REGEX = /([\u0023-\u0039][\u20e3]|[\ud800-\udbff][\udc00-\udfff]|[\u00a9-\u00ae]|[\u2122-\u3299])*([:;][^:]*)/;
+
+/**
+ * Interface to customize the emoji plugin
+ */
+export interface EmojiPluginOptions {
+    searchPlaceholder?: string;
+    searchInputAriaLabel?: string;
+}
+
+class EmojiPlugin implements ReactEditorPlugin {
+    private editor: Editor;
+    private contentDiv: HTMLDivElement;
+    private eventHandledOnKeyDown: boolean;
+    private canUndoEmoji: boolean;
+    private isSuggesting: boolean;
+    private contentEditable: HTMLDivElement;
+    private pane: EmojiPane;
+    private timer: number;
+    private uiUtilities: UIUtilities | null = null;
+    private strings: Strings;
+    private refreshCalloutDebounced: () => void;
+
+    constructor(private options: EmojiPluginOptions = {}) {
+        if (options) {
+            this.options.searchInputAriaLabel = options.searchInputAriaLabel;
+            this.options.searchPlaceholder = options.searchPlaceholder;
+        }
+
+        this.strings = {
+            ...EmojiDescriptionStrings,
+            ...EmojiKeywordStrings,
+            ...EmojiFamilyStrings,
+        };
+    }
+
+    setUIUtilities(uiUtilities: UIUtilities) {
+        this.uiUtilities = uiUtilities;
+    }
+
+    willHandleEventExclusively?: (event: PluginEvent) => boolean;
+
+    public getName() {
+        return 'Emoji';
+    }
+
+    public dispose() {
+        this.setIsSuggesting(false);
+        if (this.contentDiv) {
+            ReactDOM.unmountComponentAtNode(this.contentDiv);
+            this.contentDiv.parentElement.removeChild(this.contentDiv);
+            this.contentDiv = null;
+        }
+        this.editor = null;
+    }
+
+    public initialize(editor: Editor): void {
+        this.editor = editor;
+        const document = editor.getDocument();
+        this.contentDiv = document.createElement('div');
+        document.body.appendChild(this.contentDiv);
+    }
+
+    public initializeContentEditable(contentEditable: HTMLDivElement): void {
+        this.contentEditable = contentEditable;
+    }
+
+    public onPluginEvent(event: PluginEvent): void {
+        const domEvent = event as PluginDomEvent;
+        const keyboardEvent = domEvent.rawEvent as KeyboardEvent;
+        if (event.eventType === PluginEventType.KeyDown) {
+            this.eventHandledOnKeyDown = false;
+            if (this.isSuggesting) {
+                this.onKeyDownSuggestingDomEvent(domEvent);
+            } else if (keyboardEvent.which === KeyCodes.backspace && this.canUndoEmoji) {
+                // If KeyDown is backspace and canUndoEmoji, call editor undo
+                this.editor.undo();
+                this.handleEventOnKeyDown(domEvent);
+                this.canUndoEmoji = false;
+            }
```

Comment:
We should use auto complete api instead. Either fix it or open a bug and add bug number here to track.